### PR TITLE
Make partio publishable on PyPi

### PR DESCRIPTION
**Pull request by alexus37**
Hi,
I started this Pr since I need the python wrapper from partio in another project but don't want to put the burden of installing on the end-user. 
The idea is to build partio inside a manylinux2014 container and upload the wheel to PyPI which makes partio pip installable (pip install partio).

However, I want to mention that I am not a python packaging expert and that there is probably more elegant solutions to this than what I did. I tried to achieve this without (minimal) changes to the CMake configs.

I will leave some comments in the code to explain my building workflow. 

@davvid I claimed the partio name on PyPi and added you as a maintainer so you can publish new versions 

What is left to do:

- [ ] test if this really works and covers all use cases as a local build
- [ ] build wheel for windows 
- [ ] build wheel for mac

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -61,9 +61,13 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 set(CMAKE_FIND_PACKAGE_RESOLVE_SYMLINKS ON)
 set(CMAKE_INSTALL_MESSAGE LAZY)
-set(CMAKE_SKIP_RPATH ON)
 set(CMAKE_THREAD_PREFER_PTHREAD ON)
 set(THREADS_PREFER_PTHREAD_FLAG ON)
+# this is use for building the python wheel
```
Comment (alexus37) [inline CMakeLists.txt:66]: The idea is to package the 
"libpartio.so.1", "partio.py",  "_partio.so", into a single wheel and therefore are unpakcing the _partio.so should look in the same directory for shared objects.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+FROM quay.io/pypa/manylinux2014_x86_64:latest
+RUN (yum install cmake swig freeglut-devel zlib-devel -y)
+ADD . / io/
+CMD (cd io && /opt/python/cp38-cp38/bin/python setup.py bdist_wheel && auditwheel repair dist/partio-1.0.0-cp38-cp38-linux_x86_64.whl)
```
Comment (alexus37) [inline Dockerfile:4]: first, we have to install the required dependencies and the call setup to build the wheel.
finally, auditwheel makes the wheel publishable to PyPi

**Review by davvid — CHANGES_REQUESTED** [review]
This is looking great, thanks again for making this happen. Please amend your commits with these fixups and we'll get these updates merged soon.

File: .gitignore [hunk 3]
```diff
@@ -22,3 +22,8 @@
 *.log
 *~
 workspace.mel
+.vscode/
```
[the remark this answers is not in the document]
Comment (alexus37) [reply .gitignore:25]: pretty cool didn't know about that.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+        if platform.system() == "Windows":
+            configure_args += []
+        elif platform.system() in {"Linux", "Darwin"}:
+            configure_args += []
+        else:
+            raise RuntimeError(f"Unsupported '{platform.system()}' platform")
+
+        # Get the absolute path to the build folder
+        build_folder = str(Path(".").absolute() / f"{self.build_temp}_{ext.name}")
```
[the remark this answers is not in the document]
Comment (alexus37) [reply setup.py:65]: replaced all the f string with the format method

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            f"-DCMAKE_INSTALL_PREFIX:PATH=build/install",
+        ]
+        build_args = ["--config", ext.cmake_build_type]
+        # Extend the configure arguments with those passed from the extension
+        configure_args += ext.cmake_configure_options
+
+        if platform.system() == "Windows":
+            configure_args += []
+        elif platform.system() in {"Linux", "Darwin"}:
+            configure_args += []
```
[the remark this answers is not in the document]
Comment (alexus37) [reply setup.py:60]: Done, but the whole if statement is a bit pointless I kept there as a placeholder if some special configs args are required for different builds.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,8 @@
+FROM quay.io/pypa/manylinux2014_x86_64:latest
+RUN (yum install cmake swig freeglut-devel zlib-devel -y)
+ADD . / io/
+CMD (cd io && \
+    /opt/python/cp36-cp36m/bin/python setup.py bdist_wheel && auditwheel repair dist/partio-1.0.0-cp36-cp36m-linux_x86_64.whl && \
+    /opt/python/cp37-cp37m/bin/python setup.py bdist_wheel && auditwheel repair dist/partio-1.0.0-cp37-cp37m-linux_x86_64.whl &&  \
+    /opt/python/cp38-cp38/bin/python setup.py bdist_wheel && auditwheel repair dist/partio-1.0.0-cp38-cp38-linux_x86_64.whl &&  \
+    /opt/python/cp39-cp39/bin/python setup.py bdist_wheel && auditwheel repair dist/partio-1.0.0-cp39-cp39-linux_x86_64.whl )
```
[the remark this answers is not in the document]
Comment (alexus37) [reply Dockerfile:8]: I am using this cmd to create wheels for all currently used python versions (3.6-3.9).

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+
+setup(
+    name="partio",
+    description="A Python wrapper for partio",
+    version="1.0.0",
+    packages=[],
+    ext_modules=[PARTIO_EXT],
+    cmdclass={"build_ext": BuildCMakeExtension, "install_lib": InstallLibs,},
+    python_requires=">=3.4",
```
[the remark this answers is not in the document]
Comment (alexus37) [reply setup.py:154]: So the earliest Python 3 version we building for is 3.6 so I replaced it with 3.6

File: CMakeLists.txt (hunk 1 above)
Comment (alexus37) [reply CMakeLists.txt:66]: I added an option to the cmake and the setup.py

File: CMakeLists.txt [hunk 8]
```diff
@@ -61,9 +61,13 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
...
+set(CMAKE_SKIP_RPATH OFF)
+set(CMAKE_BUILD_RPATH "$ORIGIN")
+set(CMAKE_INSTALL_RPATH "$ORIGIN")
```
[the remark this answers is not in the document]
Comment (alexus37) [reply CMakeLists.txt:69]: I am not sure that is required in all online sources I saw they don't do it.

**Comment by alexus37** [comment]
Thanks for the detailed feedback I incorporated most of the comments.
Regarding the support of python 2.7, there are a few things that make it annoying.

First, we would need to build partio inside the manylinux1 (CentOS 5 based - EOL) docker container
which is only supported until January 2022 additionally Python 2 will reach the end of life (EOL) on January 1st, 2020.

However, I still tried to build it, but it failed, my guess is the swig version for centos is pretty old.
If you want to take a look you can just change the docker file to use quay.io/pypa/manylinux1_x86_64 as a base image
keep in mind that CMake (v2) installed on that image is too old to install partio.

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        # Everything under self.build_dir will get moved into site-packages
+        # so move all things we want installed there
+        lib_dir = self.build_dir
+        write_dir = Path(lib_dir) / "partio"
+
+        os.makedirs(write_dir, exist_ok=True)
+
+        # copy equired files
+        python_version = f"python{sys.version_info[0]}.{sys.version_info[1]}"
+        base_path = Path.cwd() / "build" / "install" / "lib64"
```
[the remark this answers is not in the document]
Comment (alexus37) [reply setup.py:126]: Not exactly sure what you mean, but I run get_conf_var on my machine I am getting the following output.

```python
In [1]: import sysconfig                                                                                                                                    
In [2]: sysconfig.get_config_var('LIBDIR')                                                                                                                  
Out[2]: '/home/ax/miniconda3/lib'


```

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        f.write("from .partio import *")
+        f.close()
+        self.announce("+++ custom done", level=3)
+        super().run()
+
+
+setup(
+    name="partio",
+    description="A Python wrapper for partio",
+    version="1.0.0",
```
[the remark this answers is not in the document]
Comment (alexus37) [reply setup.py]: I updated the version, I can take a look at the setuptools_scm tomorrow.

[the remark this answers is not in the document]
Comment (alexus37) [reply setup.py]: I included the [setuptools_scm](https://github.com/pypa/setuptools_scm) lib and added some settings in the setup.py to determine the version number. Please verify that these are correct. To find the computed version you can run: 

```python
python setup.py --version
```

**Review by davvid — APPROVED** [review]

**Comment by davvid** [comment]
Thanks again for all your help. I'll review the notes in the README.md afresh next week and attempt to do a publish into pypi. Awesome!

**Pull request merged**